tf2_py/src/tf2_py.cpp

```diff
@@ -1007,59 +1007,29 @@ static PyMethodDef module_methods[] = {
-  sizeof(buffer_core_t),                    /* tp_basicsize */
-  0,                                        /* tp_itemsize */
-  nullptr,                                     /* tp_dealloc */
-  nullptr,                                     /* tp_print */
-  nullptr,                                     /* tp_getattr */
-  nullptr,                                     /* tp_setattr */
-  nullptr,                                     /* tp_as_async */
-  nullptr,                                     /* tp_repr */
-  nullptr,                                     /* tp_as_number */
-  nullptr,                                     /* tp_as_sequence */
-  nullptr,                                     /* tp_as_mapping */
-  nullptr,                                     /* tp_hash */
-  nullptr,                                     /* tp_call */
-  nullptr,                                     /* tp_str */
-  nullptr,                                     /* tp_getattro */
-  nullptr,                                     /* tp_setattro */
-  nullptr,                                     /* tp_as_buffer */
-  Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE, /* tp_flags */
-  nullptr,                                     /* tp_doc */
-  nullptr,                                     /* tp_traverse */
-  nullptr,                                     /* tp_clear */
-  nullptr,                                     /* tp_richcompare */
-  0,                                        /* tp_weaklistoffset */
-  nullptr,                                     /* tp_iter */
-  nullptr,                                     /* tp_iternext */
-  buffer_core_methods,                      /* tp_methods */
-  nullptr,                                     /* tp_members */
-  nullptr,                                     /* tp_getset */
-  nullptr,                                     /* tp_base */
-  nullptr,                                     /* tp_dict */
-  nullptr,                                     /* tp_descr_get */
-  nullptr,                                     /* tp_descr_set */
-  0,                                        /* tp_dictoffset */
-  BufferCore_init,                          /* tp_init */
-  PyType_GenericAlloc,                      /* tp_alloc */
-  PyType_GenericNew,                        /* tp_new */
-  nullptr,                                     /* tp_free */
-  nullptr,                                     /* tp_is_gc */
-  nullptr,                                     /* tp_bases */
-  nullptr,                                     /* tp_mro */
-  nullptr,                                     /* tp_cache */
-  nullptr,                                     /* tp_subclasses */
-  nullptr,                                     /* tp_weaklist */
-  nullptr,                                     /* tp_del */
-  0,                                        /* tp_version_tag */
-  nullptr,                                     /* tp_finalize */
```

Comment:
Hm.  I'm fairly certain I did this in an earlier PR to get rid of some problem/error, though I can't remember exactly what anymore.  Any particular reason to get rid of all of this?  If you have to move some initialization out of here, I'd rather see those fields set to `nullptr` in this initialization, and then set as appropriate in `staticInit`.

Reply:
The definition of this `struct` has sometimes changed, and it's currently impossible to not get a build warning if we want to support both Python 3.7 and Python 3.8 (without adding some extra preprocessor logic).

Python 3.7: https://github.com/python/cpython/blob/50ce89071d212e659818825522be48acda4d2979/Include/object.h#L425
Python 3.8: https://github.com/python/cpython/blob/9ce361d3bb15cf49b82fa03e3e593d7cbd8ee1ff/Include/cpython/object.h#L257-L260
master: https://github.com/python/cpython/blob/fd5116c0e77aec05f67fb24f10562ac920648035/Include/cpython/object.h#L273

I think that initializing all member just makes this harder to maintain, without any extra benefit.
The non initialized members will be initialized to 0.
IMO, getting rid of the warning is safe.

Reply:
MSVC has added support of designated initializers in cpp from VS2019 v16.1 (see [here](https://developercommunity.visualstudio.com/idea/518768/c-designated-initializers.html)).
We could also use them, as all the compilers we're using have support of that feature.

Reply:
Ah yes, I was trying to get rid of `-Wmissing-field-initializers`, which is why we ended up specifying all of the fields.

Well, that's annoying.

@ivanpauno Instead of disabling the warning, what if we did something like:

```
static PyTypeObject buffer_core_Type{};
```

And then filled in *all* of the necessary fields during `staticInit`?  Do you think that will work?

(I'd rather not use designated initializers, as they are still a non-standard extension)

Reply:
> And then filled in all of the necessary fields during staticInit? Do you think that will work?

I think that will work, but the problem is how to replace `PyVarObject_HEAD_INIT`.
Based on cpython implementation:
https://github.com/python/cpython/blob/d905df766c367c350f20c46ccd99d4da19ed57d8/Include/object.h#L82-L117

I can probably do:
```cpp
buffer_core_Type.ob_base._ob_next = nullptr;
buffer_core_Type.ob_base._ob_prev = nullptr;
buffer_core_Type.ob_base.ob_refcnt = 1;
buffer_core_Type.ob_base.ob_type = nullptr;
buffer_core_Type.ob_size = 0;
```

or also:
```cpp
PyVarOvject null_var_object = PyVarObject_HEAD_INIT(nullptr, 0);
std::memcpy(&buffer_core_Type, &null_var_object, sizeof(PyVarObject));
```

But IMO, both options are just worst that ignoring the warning.

Reply:
I agree that the first one is not good; that will probably break in later versions of Python.  The version using `std::memcpy` isn't too bad, in my opinion.  But overall, I can buy your argument about disabling the warning around the structure.  So I'm OK with leaving it as-is in this patch.